faiss/impl/ProductAdditiveQuantizer.h

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <vector>
+
+#include <faiss/Index.h>
+#include <faiss/impl/AdditiveQuantizer.h>
+#include <faiss/impl/LocalSearchQuantizer.h>
+#include <faiss/impl/ResidualQuantizer.h>
+
+namespace faiss {
+
+/** Product Additive Quantizers
+ *
+ * The product additive quantizer is a variant of AQ and PQ.
+ * It first splits the vector space into multiple orthogonal sub-spaces
+ * just like PQ does. And then it quantizes each sub-space by an independent
+ * additive quantizer.
+ *
+ */
+struct ProductAdditiveQuantizer : AdditiveQuantizer {
+    size_t nsplits; ///< number of sub-vectors we split a vector into
+
+    std::vector<AdditiveQuantizer*> quantizers;
+
+    /** Construct a product additive quantizer.
+     *
+     * The ownership of the additive quantizers passed in is not
+     * transferred to the ProductAdditiveQuantizer object. Users
+     * are responsible to take care of it.
```

Comment:
This is problematic because the object cannot be self-contained in a C++ context (for example in an index_factory). 
To copy the object, we need a clone_quantizer function and a common ancestor class for the quantizer objects. 
IMO as long as we don't have the Index object implementation we don't need it, but to keep in mind for later.

Reply:
We have `ProductLocalSearchQuantizer` and `ProductResidualQuantizer` which are able to be self-contained. Should we need to make `ProductAdditiveQuantizer` self-contained too?